### ISSUE TITLE
Error "Index was outside the bounds of the array." when reading from dataset without AsNoTracking()

### ISSUE THREAD

**Issue by luizfbicalho** [dotnet/efcore#30764]
## Ask a question
I'm updating a model from EF core 5.0 to 7.0, and I have one DbContextHealthCheck that reads all datasets and reads 10 objects just to ensure that everything is working.

For one entity I'm getting the Index was outside the bounds of the array. Exception

If I use the AsNoTracking() in the query it runs fine, but I didn't find any reason for the error
I created one small project with similar entities but it works fine.

My real question is: 

How can I see or override this code to inspect more what's going on.
The problem is inside the InternalEntityEntry Constructor and in the SnapshotFactoryFactory class

```C#
 public class DbContextsHealthCheck : IHealthCheck
    {
        public DbContextsHealthCheck(IEnumerable<DbContext> contexts)
        {
            Contexts = contexts;

        }

        public IEnumerable<DbContext> Contexts { get; }

        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
        {
            foreach (var db in Contexts)
            {
                try
                {
                    await db.Database.MigrateAsync(cancellationToken);
                    var prop = db.GetType().GetProperties().Where(x => typeof(IQueryable).IsAssignableFrom(x.PropertyType));
                    foreach (var set in prop)
                    {
                        var dbset = set.GetValue(db);
                        if (dbset is IQueryable enumerable)
                        {
                            try
                            {
                                await enumerable.OfType<object>().Take(10).ToListAsync(cancellationToken);
                            }
                            catch (Exception ex)
                            {
                                return await Task.FromResult(HealthCheckResult.Degraded($"The Database {db.Database.GetDbConnection().ConnectionString}  and property {set.Name} is degraded ", ex));
                            }

                        }
                    }
                }
                catch (Exception ex)
                {
                    return await Task.FromResult(HealthCheckResult.Unhealthy($"The Database {db.Database.GetDbConnection().ConnectionString} is unhealthy ", ex));
                }
            }

            return await Task.FromResult(HealthCheckResult.Healthy("The Database is healthy"));
        }
    }
```

### Include stack traces

```
   at lambda_method339(Closure, ValueBuffer)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry..ctor(IStateManager stateManager, IEntityType entityType, Object entity, ValueBuffer& valueBuffer)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.StartTrackingFromQuery(IEntityType baseEntityType, Object entity, ValueBuffer& valueBuffer)
   at Microsoft.EntityFrameworkCore.Query.QueryContext.StartTracking(IEntityType entityType, Object entity, ValueBuffer valueBuffer)
   at lambda_method1140(Closure, QueryContext, DbDataReader, ResultContext, SingleQueryResultCoordinator)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at MC.Webapi.Extensions.HealthCheck.DbContextsHealthCheck.CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken)",
```

### Include provider and version information

EF Core version: 7.0.5
Database provider:  Microsoft.EntityFrameworkCore.SqlServer
Target framework:  .NET 7.0
Operating system: Windows 10
IDE:  Visual Studio 2022 17.6 Preview 5

**Comment by ajcvickers** [comment]
@luizfbicalho This looks like a case where the same context instance is being used by multiple threads at the same time.

**Comment by luizfbicalho** [comment]
It's surely not the case, I used it in the application startup to make sure

```C#
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            var list = app.ApplicationServices.GetRequiredService<Microsoft.EntityFrameworkCore.DbContext>().Set<Empreendimento>().ToList();

}
```

It only happens with this Entity, now I'm trying to debug this and see whats going on

**Comment by luizfbicalho** [comment]
I finally could reproduce it, this project make the error when you run it @ajcvickers 

[https://github.com/luizfbicalho/ErroModelEF7](url)

**Comment by ajcvickers** [comment]
@luizfbicalho I'm still not able to reproduce this. Does it need data in the database?

**Comment by luizfbicalho** [comment]
Sorry @ajcvickers 

Updated the version, created a new database and added data to it

**Comment by ajcvickers** [comment]
Note for triage: looks like a bug in unidirectional many-to-many relationships. Minimal repro below.

@luizfbicalho I was able to work around this by adding a navigation property between `Autorizacao` and `Perfil`:

```C#
public virtual IList<Perfil> Perfils { get; set; } = new List<Perfil>();
```

And using it in the many-to-many relationship:

```C#
modelBuilder.Entity<Perfil>().HasMany(t => t.Autorizacoes).WithMany(e => e.Perfils)
```

Repro:

```C#
using (var context = new SomeDbContext())
{
    await context.Database.EnsureDeletedAsync();
    await context.Database.EnsureCreatedAsync();

    var emp = new Empreendimento { Praca = new Praca() };
    context.Add(emp);
    context.Add(emp.Praca);
    context.SaveChanges();

    await context.SaveChangesAsync();
}

using (var context = new SomeDbContext())
{
    var emps = context.Empreendimento.ToList();
}

public class Autorizacao
{
    public virtual int Id { get; set; }
}

public class Empreendimento : Autorizacao
{
    public virtual int Id { get; set; }
    public virtual Praca? Praca { get; set; }
}

public class Perfil
{
    public virtual int Id { get; set; }
    public virtual IList<Autorizacao> Autorizacoes { get; set; } = new List<Autorizacao>();
}

public class Praca
{
    public virtual int Id { get; set; }
}

public class SomeDbContext : DbContext
{
    public DbSet<Empreendimento> Empreendimento => Set<Empreendimento>();
    public DbSet<Autorizacao> Autorizacao => Set<Autorizacao>();

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder
            .UseSqlServer(@"Data Source=(LocalDb)\MSSQLLocalDB;Database=AllTogetherNow")
            .LogTo(Console.WriteLine, LogLevel.Information)
            .EnableSensitiveDataLogging();
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Perfil>().HasMany(t => t.Autorizacoes).WithMany();
    }
}
```

**Comment by luizfbicalho** [comment]
Thanks @ajcvickers  for a workaround, I'll try and see if it doesn't break anything else.

Do you think this will work on EF 7 or only in EF 8?

**Comment by luizfbicalho** [comment]
I updated to the 7.0.8 and it didn't fix
and using the latest version of the 8.0 preview was working

**Comment by ajcvickers** [comment]
Looks like it didn't make 7.0.8 and will be included in 7.0.9.
